:memo: Rework PHP-FPM sizing (#2702)
platformsh/platformsh-docs: docs/src/languages/php/fpm.md

By default, Platform.sh automatically sets a maximum number of PHP-FPM workers for your app. 
This number is calculated based on three parameters:
- The amount of memory you can allot for PHP processing depending on your server specifications.
- The request memory.
  This is the amount of memory an average PHP request is expected to require.
- The reserved memory. 
  This is the amount of memory you need to reserve for tasks that aren't related to requests. 

## The heuristic
Follow the steps to adjust the maximum number of PHP-FPM workers depending on your app's needs.

The heuristic is based on three input parameters:
## Before you begin

 * The memory available for the container, which depends on the size of the container (`S`, `M`, `L`),
 * The memory that an average request is expected to require,
 * The memory that should be reserved for things that aren't specific to a request (memory for `nginx`, the op-code cache, some OS page cache, etc.)
You need:

The number of workers is calculated as:
- An up-and-running web app in PHP, complete with [PHP-FPM](https://www.php.net/manual/en/install.fpm.php)
- The [Platform.sh CLI](../../administration/cli/_index.md)

> ![FPM](/images/php/phpfpmworkers.png "0.3")

Author: I think a formula like this can sometimes be better than words and bring a lot of clarity. But in this case, I removed it because it actually made everything more confusing to me, and I'm still not sure I completely understand what each element in it means. Granted, I'm terrible at maths, but I figured if it made things more obscure for me, then it might very well be the case for other people too.

> ![FPM](/images/php/phpfpmworkers.png "0.3")

Reviewer: I looked into it some more and the `max` and `2` bits mean that it sets a minimum value of 2 (so if the other part of the formula comes out to less than 2, the result is 2). That information is missing elsewhere, but is definitely not ideally presented in the image (let alone only being in image form).

The other three parts of the formula are the three bullet points above. So it should be possible to make them clearer. I think seeing that RequestMemory is on the bottom of the equation is helpful to seeing quickly that it's the inverse -- so increasing it decreases the number of workers.

I also discovered that the result is probably rounded up when it's a fraction. Not sure how much that info helps, but it might.

So I'd replace it with just the three items and include the minimum elsewhere in the text. At the very least, including the minimum.

---
title: "PHP-FPM sizing"
description: "Learn how to adjust the maximum number of PHP-FPM workers for your app"
weight: 5
---

PHP-FPM helps improve your app's performance
by maintaining pools of workers that can process PHP requests.
This is particularly useful when your app needs to handle a high number of simultaneous requests. 

By default, Platform.sh automatically sets a maximum number of PHP-FPM workers for your app. 
This number is calculated based on three parameters:
- The container memory: the amount of memory you can allot for PHP processing 
  depending on [app size](../../create-apps/app-reference.md#sizes).
- The request memory: the amount of memory an average PHP request is expected to require.
- The reserved memory: the amount of memory you need to reserve for tasks that aren't related to requests. 

> The calculation is performed as follows: ![FPM](/images/php/PHP-FPM-Workers-Calculation.png "0.2")

Reviewer: :hammer_and_wrench: I'm somewhat allergic to the phrasing `is performed` as it usually seems to me to involve extra words. One of the two suggestions might fit?

Also added meaningful alt text.
```suggestion
The number is calculated as follows: ![The sum of container memory minus reserved memory divided by request memory](/images/php/PHP-FPM-Workers-Calculation.png "0.2")
```

Reviewer: ```suggestion
This is calculated as follows: ![The sum of container memory minus reserved memory divided by request memory](/images/php/PHP-FPM-Workers-Calculation.png "0.2")
```

Author: Didn't mean to trigger you! 😁 But I get it!

Note that when the resulting number is a decimal, 

> it is rounded up to set the maximum number of workers.

Reviewer: :pick:  As noted by the linting, but definitely not a blocker
```suggestion
it's rounded up to set the maximum number of workers.
```

Also, the minimum number of PHP-FPM workers is 2.  

To adjust the maximum number of PHP-FPM workers depending on your app's needs, follow the instructions on this page.

## Before you begin

You need:

- An up-and-running web app in PHP, complete with [PHP-FPM](https://www.php.net/manual/en/install.fpm.php)
- The [Platform.sh CLI](../../administration/cli/_index.md)

Note that the memory settings mentioned on this page are different from the [`memory_limit` PHP setting](./_index.md). 
The `memory_limit` setting is the maximum amount of memory a single PHP process can use 

> before it is automatically terminated.

Reviewer: :pick:  As noted by the linting, but definitely not a blocker
```suggestion
before it's automatically terminated.
```

```yaml {location=".platform.app.yaml"}
runtime:
    sizing_hints:
        request_memory: 110
```

Note that the minimum value for the `request_memory` key is 10 MB
and the minimum value for the `reserved_memory` key is 70 MB.
If you set lower values, they are automatically overridden with those minimums.

For information on how to estimate the best `request_memory` value depending on your needs, 
see how to [measure PHP worker memory usage](./fpm.md#measure-php-worker-memory-usage). 

To check the maximum number of PHP-FPM workers available to your site,
open an [SSH session](/development/ssh/_index.md) and run the following command:

```bash
grep -e '^pm.max_children' /etc/php/*/fpm/php-fpm.conf
```

You get output similar to the following,
where workers are referred to as `children`:

```bash
pm.max_children = 2
```

## Measure PHP worker memory usage

To see how much memory your PHP worker processes use, 
open an [SSH session](/development/ssh/_index.md) 
and look at the PHP access log:

```bash
less /var/log/php.access.log
```

In the fifth column of the output, 
you can see the peak memory usage that occurred when each request was handled. 

> The optimal request memory is somewhere between the average and worst case memory usages listed.

Reviewer: :lady_beetle: `somewhere between` seems rather vague. I'm not sure how useful it really is.

To help determine what the optimal request memory is, run the following command:

```bash
tail -n5000 /var/log/php.access.log | awk '{print $6}' | sort -n | uniq -c
```

> Note that the above command takes into account the last 5000 requests that reached PHP-FPM.
> You can adjust this number depending on the amount of traffic on your site.

Reviewer: :+1: :hammer_and_wrench: This is more clear that the original, for sure. We do try to avoid directional language like `above`.
```suggestion
This command takes into account the last 5,000 requests that reached PHP-FPM.
You can adjust this number depending on the amount of traffic on your site.
```

In the output, you can see how many requests used how much memory, in KB.
For example:

```text

>       1

Reviewer: :question: Why is there one request without a memory log? I think it should be removed, but I don't understand it.

      6 131072
```

In the above output:

> - The majority of requests (4800) peaked at 2048 KB of memory, 
>   likely thanks to application caching.
> - Most other requests (3206) used up to around 10 MB of memory.

Reviewer: :+1: Clearer as a list, for sure.

:clipboard: https://developers.google.com/style/numbers#commas-and-decimal-points-in-numbers
```suggestion
- The majority of requests (4,800) peaked at 2,048 KB of memory, 
  likely thanks to app caching.
- Most other requests (3,206) used up to around 10 MB of memory.
```

> - 758 requests used more than 10 MB and up to around 18 MB of memory.

Reviewer: :lady_beetle: The numbers `3,206` and `758` aren't very clear from the example. I'd probably simplify the example to make it easier to see the example distinctions. Like there's one at 22.5 MB and where does that fit? Getting rid of some might help.

- Only 6 requests peaked at 131 MB of memory, 
  likely due to cache clears.

In such a situation, you might want to remain cautious and set an average request memory of 16 MB.

> A more aggressive approach might be for you to set an average request memory of 10 MB instead.

Reviewer: :nail_care: 
```suggestion
A more aggressive approach might be for you to set an average request memory of 10 MB.
```

> Note that, with the more aggressive approach, there is a risk of allowing more concurrent requests 
> and needing to use swap memory. 

Reviewer: :pick: To try to for more directness and fewer words
```suggestion
With the more aggressive approach, you a risk of allowing more concurrent requests and needing to use swap memory. 
```

Although the idea of `swap memory` isn't clear to everyone and may be more confusing than helpful. :shrug:

This can result in latencies.

For further help in estimating your optimal average request memory, 
you can use the [log analyzer tool for Platform.sh](https://github.com/pixelant/platformsh-analytics) 
from the web agency [Pixelant](https://www.pixelant.net/). 
This tool offers a better visualization of access logs to determine how much memory requests use on average.
It also offers additional insights into the operation of your site. 
These can help further optimize your configuration 
and provide guidance on when to increase your plan size.
Note that this tool is maintained by a third party, not by Platform.sh.

> For information on how to set an average request memory, see how to [adjust default assumptions](./fpm.md#adjust-default-assumptions).

Reviewer: :clipboard: 
```suggestion
For information on how to set the average request memory, see how to [adjust the default assumptions](#adjust-default-assumptions).
```
